auslib/test/test_AUS.py

```diff
@@ -3,56 +3,49 @@
 import unittest
 
 from auslib.global_state import dbo
-from auslib.AUS import AUS
+from auslib.AUS import AUS, AUSRandom, SUCCEED, FAIL
 from auslib.blobs.base import createBlob
 
 
+def getRange(rand):
+    return range(rand.min, rand.max + 1)
+
+
 def setUpModule():
     # Silence SQLAlchemy-Migrate's debugging logger
     logging.getLogger('migrate').setLevel(logging.CRITICAL)
 
 
-def RandomAUSTestWithoutFallback(AUS, backgroundRate, force, mapping):
-    with mock.patch('auslib.db.Rules.getRulesMatchingQuery') as m:
-        m.return_value = [dict(backgroundRate=backgroundRate, priority=1, mapping=mapping, update_type='minor',
-                               fallbackMapping=None)]
+class TestAUSThrottlingWithoutFallback(unittest.TestCase):
 
-        results = AUS.rand.getRange()
-        resultsLength = len(results)
+    def setUp(self):
+        dbo.setDb('sqlite:///:memory:')
+        dbo.create()
+        dbo.releases.t.insert().execute(
+            name='b', product='b', data_version=1,
+            data=createBlob({"name": "b", "extv": "1.0", "schema_version": 1, "platforms": {"a": {"buildID": "1", "locales": {"a": {}}}}}))
 
-        def se(*args, **kwargs):
-            return results.pop()
-        with mock.patch('auslib.AUS.AUSRandom.getInt') as m2:
-            m2.side_effect = se
-            served = 0
-            tested = 0
-            while len(results) > 0:
-                updateQuery = dict(
-                    channel='foo', force=force, buildTarget='a', buildID='0',
-                    locale='a', version='1.0',
-                )
-                r, _ = AUS.evaluateRules(updateQuery)
-                tested += 1
-                if r:
-                    served += 1
-                # bail out if we're not asking for any randint's
-                if resultsLength == len(results):
-                    break
-            return (served, tested)
+        dbo.releases.t.insert().execute(
+            name='fallback', product='c', data_version=1,
+            data=createBlob({"name": "fallback", "extv": "1.0", "schema_version": 1, "platforms": {"a": {"buildID": "1", "locales": {"a": {}}}}}))
+
+    def tearDown(self):
+        dbo.reset()
 
+    def random_aus_test(self, background_rate, force=None, fallback=False):
+        mapping = 'b'
+        with mock.patch('auslib.db.Rules.getRulesMatchingQuery') as m:
+            fallback = fallback and 'fallback'  # convert True to string
```

Comment:
I'm not sure I understand this line. The string 'fallback' will always evaluate to True, so it seems like this is the same as bool(fallback)? If so, I don't think this will set the value correctly. The returned value from the database should be the name of the fallback release ("fallback" in these tests) if present, or None if there isn't a fallbackMapping specified.

I could be misunderstanding something here, so feel free to correct me.

Reply:
Another way to write this line of code is

```python
if fallback:
    fallback = 'fallback'
```

In other words, the `and` operator takes the first falsy value, or the last value if all are truthy. Some examples:

```python
>>> 'abc' and 'def'
'def'
>>> "" and 'def'
""
>>> False and 'def'
False
>>> True and 'def'
'def'
```

Here, the point is to convert `True` (meaning "yes, turn on `fallback`, please") becomes the string `'fallback'`, since `True` is more convenient at the API layer but not a valid fallback release (since it has to be a name, i.e. string).

Reply:
Huh, TIL. Thanks for the detailed explanation.